rfd/0180-sso-as-mfa-method.md

```diff
@@ -0,0 +1,767 @@
+
+```proto
+message OIDCConnectorMFASettings {
+  // Enabled specified whether this SAML connector supports MFA checks. Defaults to false.
+  bool enabled = 1;
+  // MFAMode specifies what MFA mode this auth connector will operate with. Defaults to "optional".
+  AuthConnectorMFAMode mfa_mode = 2;
+  // ClientID is the id of the authentication client (Teleport Auth server). If unset, 
+  // the parent OIDC connector's ClientID will be used.
+  string client_id = 3;
+  // ClientSecret is used to authenticate the client. If unset, the parent OIDC connector's
+  // ClientSecret will be used.
+  string client_secret = 4;
+  // AcrValues are Authentication Context Class Reference values. The meaning of the ACR
+  // value is context-specific and varies for identity providers. Some identity providers
+  // support MFA specific contexts, such Okta with its "phr" (phishing-resistant) ACR.
+  string acr_values = 5;
+  // Prompt is an optional OIDC prompt. An empty string omits prompt.
+  // If not specified, it defaults to select_account for backwards compatibility.
+  string prompt = 6;
+  // MaxAge is the amount of time that an IdP session is valid for. Defaults to 0 to always
+  // force re-authentication for MFA checks. This should only be set to a non-zero value if
+  // the IdP is setup to perform MFA checks on top of active user sessions.
+  google.protobuf.Duration max_age = 67;
+}
+```
+
+```diff
+message OIDCConnectorSpecV3 {
+  ...
++ // MFASettings contains settings to enabled SSO MFA checks through this auth connector.
++ OIDCConnectorMFASettings MFASettings = 19 [(gogoproto.jsontag) = "mfa_settings,omitempty"];
+}
+```
+
+**GithubConnectorMFASettings**
+
+```proto
+message GithubConnectorMFASettings {
+  // Enabled specified whether this SAML connector supports MFA checks. Defaults to false.
+  bool enabled = 1;
+  // MFAMode specifies what MFA mode this auth connector will operate with. Defaults to "optional".
+  AuthConnectorMFAMode mfa_mode = 2;
+  // ClientID is the id of the authentication client (Teleport Auth server). If unset, the parent 
+  // Github connector's ClientID will be used.
+  string client_id = 3;
+  // ClientSecret is used to authenticate the client. If unset, the parent Github connector's
+  // ClientSecret will be used.
+  string client_secret = 4;
+  // MaxAge is the amount of time that an IdP session is valid for. Defaults to 0 to always
+  // force re-authentication for MFA checks. This should only be set to a non-zero value if
+  // the IdP is setup to perform MFA checks on top of active user sessions.
+  google.protobuf.Duration max_age = 5;
+}
+```
+
+```diff
+message GithubConnectorSpecV3 {
+  ...
++ // MFASettings contains settings to enabled SSO MFA checks through this auth connector.
++ GithubConnectorMFASettings MFASettings = 10 [(gogoproto.jsontag) = "mfa_settings,omitempty"];
+}
+```
+
+### Backward Compatibility
+
+The SSO MFA flow is completely separate from the SSO login flow, meaning a new
```

Comment:
If they are just doing `tsh login` no. If they are doing a "retry-with-relogin" with `tsh ssh`, then yes they will have to go through the login flow and then the MFA flow. In the intended use case this would simply be a repeat of the exact same authentication.

Note that this is exactly like webauthn: we prompt them to once to tap for login, and again to connect to the node.

Maybe your question was if they need to do the SSO MFA flow to login in the first place? SSO MFA will not be used for login, it is only used for MFA protected actions within a Teleport login session (per-session MFA, admin actions, moderated session, etc.)